observer/probers/tls/tls.go

```diff
@@ -0,0 +1,193 @@
+package probers
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"golang.org/x/crypto/ocsp"
+)
+
+type reason int
+
+type badOutcomeError struct {
+	reason reason
+}
+
+const (
+	none reason = iota
+	internalError
+	issuerVerifyFailed
+	ocspUnknown
+	ocspError
+)
+
+var reasonToString = map[reason]string{
+	none:               "nil",
+	internalError:      "internalError",
+	issuerVerifyFailed: "issuerVerifyFailed",
+	ocspUnknown:        "ocspUnknown",
+	ocspError:          "ocspError",
+}
+
+func (e badOutcomeError) Error() string {
+	return reasonToString[e.reason]
+}
+
+func getReasons() []string {
+	var allReasons []string
+	for _, v := range reasonToString {
+		allReasons = append(allReasons, v)
+	}
+	return allReasons
+}
+
+// TLSProbe is the exported `Prober` object for monitors configured to perform
+// TLS protocols.
+type TLSProbe struct {
+	url        string
+	rootOrg    string
+	rootCN     string
+	response   string
+	certExpiry *prometheus.GaugeVec
+	reason     *prometheus.GaugeVec
+}
+
+// Name returns a string that uniquely identifies the monitor.
+func (p TLSProbe) Name() string {
+	return fmt.Sprintf("%s-expecting-%s-%s-%s", p.url, p.response, p.rootOrg, p.rootCN)
```

Comment:
In theory, this name only needs to contain the URL -- it would be very odd to have two probers both pointed at the *same* URL but expecting different results. Maybe so odd that we want to prevent it, by having their names collide if someone tries to configure it that way.